File: react-ui/packages/web/src/shared/enhancedEnums/userRedemptionStatus.ts
import { UserRedemptionStatus } from '@narrative/shared';
import { EnumEnhancer } from '../utils/enhancedEnumUtils';

/**
 * jw: This enum is closely tied to the java version of this enum, particularly since it duplicates the walletUpdatable
 *     and supportsRedemption flags. It's vital that this is kept in sync with the Java version.
 */

// jw: let's define the QualityFilterHelper that will provide all the extra helper logic for QualityFilters
export class UserRedemptionStatusHelper {
  status: UserRedemptionStatus;
  walletUpdatable: boolean;
  supportsRedemption: boolean;

  constructor(
    status: UserRedemptionStatus,
    walletUpdatable: boolean,
    supportsRedemption: boolean
  ) {
    this.status = status;
    this.walletUpdatable = walletUpdatable;
    this.supportsRedemption = supportsRedemption;
  }

  public isWalletUnspecified(): boolean {
    return this.status === UserRedemptionStatus.WALLET_UNSPECIFIED;
  }

  public isWalletInWaitingPeriod(): boolean {
    return this.status === UserRedemptionStatus.WALLET_IN_WAITING_PERIOD;
  }

  public isHasPendingRedemption(): boolean {
    return this.status === UserRedemptionStatus.HAS_PENDING_REDEMPTION;
  }

  public isRedemptionAvailable(): boolean {
    return this.status === UserRedemptionStatus.REDEMPTION_AVAILABLE;
  }
}

// jw: next: lets create the lookup of QualityFilter to helper object

const helpers: {[key: number]: UserRedemptionStatusHelper} = [];
helpers[UserRedemptionStatus.WALLET_UNSPECIFIED] = new UserRedemptionStatusHelper(
  UserRedemptionStatus.WALLET_UNSPECIFIED,
  true,
  false
);
helpers[UserRedemptionStatus.WALLET_IN_WAITING_PERIOD] = new UserRedemptionStatusHelper(
  UserRedemptionStatus.WALLET_IN_WAITING_PERIOD,
  true,
  false
);
helpers[UserRedemptionStatus.HAS_PENDING_REDEMPTION] = new UserRedemptionStatusHelper(
  UserRedemptionStatus.HAS_PENDING_REDEMPTION,
  false,
  false
);
helpers[UserRedemptionStatus.REDEMPTION_AVAILABLE] = new UserRedemptionStatusHelper(
  UserRedemptionStatus.REDEMPTION_AVAILABLE,
  true,
  true
);

// jw: finally, let's create the enhancer, which will allow us to lookup helpers by enum instance.
export const EnhancedUserRedemptionStatus = new EnumEnhancer<UserRedemptionStatus, UserRedemptionStatusHelper>(
  helpers
);

File: core/src/main/java/org/narrative/network/core/user/UserRedemptionStatus.java
package org.narrative.network.core.user;

import org.narrative.common.util.enums.IntegerEnum;

/**
 * Date: 2019-06-26
 * Time: 13:32
 *
 * @author jonmark
 */
public enum UserRedemptionStatus implements IntegerEnum {
    // jw: This enum is closely tied to the React enhanced version (userRedemptionStatus.ts) of this enum, particularly
    //     since the walletUpdatable and supportsRedemption flags are duplicated there. It's vital that changes made here
    //     are reflected in the enhanced typescript version as well.
    WALLET_UNSPECIFIED(0, true, false)
    ,WALLET_IN_WAITING_PERIOD(1, true, false)
    ,HAS_PENDING_REDEMPTION(2, false, false)
    ,REDEMPTION_AVAILABLE(4, true, true)
    ;

    private final int id;
    private final boolean walletUpdatable;
    private final boolean supportsRedemption;

    UserRedemptionStatus(int id, boolean walletUpdatable, boolean supportsRedemption) {
        this.id = id;
        this.walletUpdatable = walletUpdatable;
        this.supportsRedemption = supportsRedemption;
    }

    @Override
    public int getId() {
        return id;
    }

    public boolean isWalletUpdatable() {
        return walletUpdatable;
    }

    public boolean isSupportsRedemption() {
        return supportsRedemption;
    }

    public boolean isWalletInWaitingPeriod() {
        return this == WALLET_IN_WAITING_PERIOD;
    }
}